lib/BlockedBy.js

```diff
@@ -2,7 +2,55 @@
 
 const NodeResque = require('node-resque');
 const helper = require('./helper.js');
-const { runningJobsPrefix, waitingJobsPrefix } = require('../config/redis');
+const hoek = require('hoek');
+const { runningJobsPrefix, waitingJobsPrefix, queuePrefix } = require('../config/redis');
+const winston = require('winston');
+
+/**
+ * collapse waiting builds and re-enequeue the current build if it is the latest one
+ * @method collapseBuilds
+ * @param  {String}  waitingKey           ${waitingJobsPrefix}${jobId}
+ * @param  {Number}  buildId              Current build Id
+ * @param  {Array}   blockingBuildIds     List of build Ids that are blocking this current build
+ */
+async function collapseBuilds({ waitingKey, buildId, blockingBuildIds }) {
```

Comment:
I think this logic might be simpler if we call it after `blockedBySelf`. That way `blockedBySelf` will do the adding to `waitingKey`, and `collapseBuilds` only has the responsibility to remove extra keys except for the last one.

Reply:
discussed offline, putting it after blockedBySelf may result in re-enqueue of the older builds